File: services/graphTransformer.js
/**
 * Created by Ivan O'Mahony on 9/14/2014.
 */
"use strict";
var GraphTransformerError = require('./graphTransformerError');
var logger = require('winston');

// This represents the server side functions to transform the client side graph (d3 node / links) to the server side
// model
function GraphTransformer() {
}
// Given a server graph transform it into a client side graph, see tests for sample data sets
GraphTransformer.prototype.toClientGraph = function (data) {
    if (typeof data === 'undefined' || data === "") {
        logger.error("toClientGraph: Triplets not defined");
        throw new GraphTransformerError("toClientGraph: Triplets not defined");
    }
    var nodes = {};
    // Create node list, this should be a unique set
    for (var i = 0; i < data.triplets.length; i++) {
        nodes[data.triplets[i][0].source.url] = data.triplets[i][0].source;
        nodes[data.triplets[i][2].target.url] = data.triplets[i][2].target;
    }
    var uniqueNodes = [];
    for (var key in nodes) {
        uniqueNodes.push({"name": nodes[key].data.name, "group": nodes[key].label, "url": nodes[key].url});
    }
    var links = [];
    // Loop relationships next
    for (var j = 0; j < data.triplets.length; j++) {
        var srcIndex = indexOfNode(uniqueNodes, data.triplets[j][0].source.url);
        var targetIndex = indexOfNode(uniqueNodes, data.triplets[j][2].target.url);
        links.push({"source": srcIndex, "target": targetIndex, "value": 1, "url": data.triplets[j][1].relationship.url});
    }
    return {"nodes": uniqueNodes, "links": links};
};

function indexOfNode(nodes, value) {
    for (var i = 0; i < nodes.length; i++) {
        if (nodes[i].url == value) {
            return i;
        }
    }
}

module.exports = GraphTransformer;

File: server/services/GraphTransformer.py
__author__ = 'Ivan'

import logging

from pydash.objects import deep_get

from server.error import GraphTransformerException

log = logging.getLogger(__name__)


class GraphTransformer(object):
    @staticmethod
    def to_client_graph(data):
        """
        Takes a given a server graph, in JSON format and transforms it into a client side graph, in JSON format.
        NOTE: The client side graph here is in a vaild D3 format of nodes and links, more details in
        public/javascripts/QueryEditor.js
        :param data: The server side, e.g. data container representation of the data
        :return: JSON data in the client format
        (http://docs.python-requests.org/en/latest/user/quickstart/#errors-and-exceptions)
        """
        if not data:
            log.error("No data to transform")
            raise GraphTransformerException("No data to transform")
        log.debug("About to parse " + str(data))
        nodes = {}
        triplets = deep_get(data, 'triplets')
        for trip in triplets:
            nodes[deep_get(trip, '0.source.url')] = deep_get(trip, '0.source')
            nodes[deep_get(trip, '2.target.url')] = deep_get(trip, '2.target')

        uniq_nodes = []
        for node in nodes.values():
            uniq_nodes.append({"name": node['data']['name'], "group": node['label'], "url": node['url']})

        links = []
        for trip in triplets:
            links.append({"source": GraphTransformer.index_of_node(uniq_nodes, deep_get(trip, '0.source.url')),
                          "target": GraphTransformer.index_of_node(uniq_nodes, deep_get(trip, '2.target.url')),
                          "value": 1,
                          "url": deep_get(trip, '1.relationship.url')})

        res = {"nodes": uniq_nodes, "links": links}
        log.debug("Returning " + str(res))
        return res

    @staticmethod
    def index_of_node(nodes, url):
        """
       Given a list of dictionaries and a url to find in a dictionary returns the index of the matching dictionary
       :param nodes: The list of nodes
       :param nodes: The node_key to find the index of
       :return: Int the index of the node_key in the list
       """
        for i, k in enumerate(nodes):
            if k['url'] == url:
                return i
